pkg/costmodel/allocation.go

```diff
@@ -549,164 +549,151 @@ func (cm *CostModel) computeAllocation(start, end time.Time, resolution time.Dur
+			cNodeKey := newNodeKey(cluster, nodeName)
 
-			node := cm.getNodePricing(nodeMap, nodeKey)
+			node := cm.getNodePricing(nodeMap, cNodeKey)
 			alloc.Properties.ProviderID = node.ProviderID
 			alloc.CPUCost = alloc.CPUCoreHours * node.CostPerCPUHr
 			alloc.RAMCost = (alloc.RAMByteHours / 1024 / 1024 / 1024) * node.CostPerRAMGiBHr
 			alloc.GPUCost = alloc.GPUHours * node.CostPerGPUHr
 
-			if pvcs, ok := podPVCMap[podKey]; ok {
-				for _, pvc := range pvcs {
+			// Make sure that the name is correct (node may not be present at this
+			// point due to it missing from queryMinutes) then insert.
+			alloc.Name = fmt.Sprintf("%s/%s/%s/%s/%s", cluster, nodeName, namespace, podName, container)
+			allocSet.Set(alloc)
+		}
+	}
 
-					pvcKey := newPVCKey(cluster, namespace, pvc.Name)
+	return allocSet, nil
+}
 
-					s, e := alloc.Start, alloc.End
-					if pvcInterval, ok := pvcPodIntervalMap[pvcKey][podKey]; ok {
-						s, e = *pvcInterval.Start(), *pvcInterval.End()
-					} else {
-						log.Warnf("CostModel.ComputeAllocation: allocation %s and PVC %s have no associated active window", alloc.Name, pvc.Name)
-					}
+func applyPVCsToPods(window kubecost.Window, podMap map[podKey]*Pod, podPVCMap map[podKey][]*PVC, pvcMap map[pvcKey]*PVC) {
+	// Because PVCs can be shared among pods, the respective PV cost
+	// needs to be evenly distributed to those pods based on time
+	// running, as well as the amount of time the PVC was shared.
 
-					minutes := e.Sub(s).Minutes()
-					hrs := minutes / 60.0
+	// Build a relation between every PVC to the pods that mount it
+	// and a window representing the interval during which they
+	// were associated.
+	pvcPodWindowMap := make(map[pvcKey]map[podKey]kubecost.Window)
+
+	for thisPodKey, pod := range podMap {
+		if pvcs, ok := podPVCMap[thisPodKey]; ok {
+			for _, pvc := range pvcs {
+
+				// Determine the (start, end) of the relationship between the
+				// given PVC and the associated Allocation so that a precise
+				// number of hours can be used to compute cumulative cost.
+				s, e := pod.Start, pod.End
+				if pvc.Start.After(pod.Start) {
+					s = pvc.Start
+				}
+				if pvc.End.Before(pod.End) {
+					e = pvc.End
+				}
 
-					count := float64(pvc.Count)
-					if pvc.Count < 1 {
-						count = 1
-					}
+				thisPVCKey := pvc.Key()
+				if pvcPodWindowMap[thisPVCKey] == nil {
+					pvcPodWindowMap[thisPVCKey] = make(map[podKey]kubecost.Window)
+				}
 
-					gib := pvc.Bytes / 1024 / 1024 / 1024
-					cost := pvc.Volume.CostPerGiBHour * gib * hrs
+				pvcPodWindowMap[thisPVCKey][thisPodKey] = kubecost.NewWindow(&s, &e)
+			}
+		}
+	}
 
-					// Scale PV cost by PVC sharing coefficient.
-					if coeffComponents, ok := sharedPVCCostCoefficientMap[pvcKey][podKey]; ok {
-						cost *= getCoefficientFromComponents(coeffComponents)
-					} else {
-						log.Warnf("CostModel.ComputeAllocation: allocation %s and PVC %s have relation but no coeff", alloc.Name, pvc.Name)
-					}
```

Comment:
This is where the total PVC cost is being multiplied by Coefficients which could sum to greater than 100%